connector/ldap/ldap.go

```diff
@@ -496,52 +503,107 @@ func (c *ldapConnector) groups(ctx context.Context, user ldap.Entry) ([]string,
+		if obtainedGroups, filter, err := c.queryGroups(ctx, attr); err == nil {
+			groups = append(groups, obtainedGroups...)
+			if len(obtainedGroups) == 0 {
+				// TODO(ericchiang): Is this going to spam the logs?
+				c.logger.Errorf("ldap: groups search with filter %q returned no groups", filter)
+			}
+		} else {
+			return nil, err
 		}
+	}
 
-		req := &ldap.SearchRequest{
-			BaseDN:     c.GroupSearch.BaseDN,
-			Filter:     filter,
-			Scope:      c.groupSearchScope,
-			Attributes: []string{c.GroupSearch.NameAttr},
-		}
+	if c.GroupSearch.Inheritance {
+		c.logger.Infof("Inheritance enabled for groups search")
+	}
+	// While loop for search in different levels (of LDAP tree) with inheritance
+	// or one time with inheritance disabled
+	// This allows to give permissions to subgroups and affect all users of subgroups
+	// Note that this iterative implementation optimizes the recursive one from LDAP
+	for {
+		// Temporal variable used to reset variable groups on each cycle
+		var nextLevelGroups []*ldap.Entry
+		for _, group := range groups {
+
+			name := getAttr(*group, c.GroupSearch.NameAttr)
+			if name == "" {
+				// Be obnoxious about missing missing attributes. If the group entry is
+				// missing its name attribute, that indicates a misconfiguration.
+				//
+				// In the future we can add configuration options to just log these errors.
+				return nil, fmt.Errorf("ldap: group entity %q missing required attribute %q",
+					group.DN, c.GroupSearch.NameAttr)
+			}
 
-		gotGroups := false
-		if err := c.do(ctx, func(conn *ldap.Conn) error {
-			c.logger.Infof("performing ldap search %s %s %s",
-				req.BaseDN, scopeString(req.Scope), req.Filter)
-			resp, err := conn.Search(req)
-			if err != nil {
-				return fmt.Errorf("ldap: search failed: %v", err)
+			// Prevent duplicates and circular hierarchy
+			exit := false
+			for _, groupName := range groupNames {
+				if name == groupName {
+					c.logger.Infof("Found duplicate group with name %s", name)
+					exit = true
+					break
+				}
+				c.logger.Infof("Comparing %s with %s", name, groupName)
+			}
+			if exit {
+				continue
+			}
+
+			groupNames = append(groupNames, name)
+			if c.GroupSearch.Inheritance {
+				if obtainedGroups, _, err := c.queryGroups(ctx, group.DN); err == nil {
+					// Keep searching upwards
+					if len(obtainedGroups) != 0 {
+						nextLevelGroups = append(nextLevelGroups, obtainedGroups...)
+					} else {
+						c.logger.Infof("Didn't find parents for group with name %s", name)
+					}
+				} else {
+					return nil, err
+				}
 			}
-			gotGroups = len(resp.Entries) != 0
-			groups = append(groups, resp.Entries...)
-			return nil
-		}); err != nil {
-			return nil, err
 		}
-		if !gotGroups {
-			// TODO(ericchiang): Is this going to spam the logs?
-			c.logger.Errorf("ldap: groups search with filter %q returned no groups", filter)
+		// If there's no remaining levels -> exit loop
+		if len(nextLevelGroups) == 0 {
```

Comment:
What prevents this from looping infinitely?

Reply:
The infinite loop is prevented here: https://github.com/paroque28/dex/blob/master/connector/ldap/ldap.go#L542
I'm going to add integration testing for this